core/chains/evm/forwarders/forwarder_manager.go

```diff
@@ -0,0 +1,299 @@
+type Config interface {
+	LogSQL() bool
+}
+
+var AuthTopics = []common.Hash{
+	authorized_receiver.AuthorizedReceiverAuthorizedSendersChanged{}.Topic(),
+	offchain_aggregator_wrapper.OffchainAggregatorConfigSet{}.Topic(),
+}
+
+var ForwardABI = evmtypes.MustGetABI(authorized_forwarder.AuthorizedForwarderABI).Methods["forward"]
+
+var AuthLogCheckInterval = time.Duration(5 * time.Second)
+
+type FwdMgr struct {
+	ORM       ORM
+	config    Config
+	evmClient evmclient.Client
+	logger    logger.Logger
+	logpoller *evmlogpoller.LogPoller
+
+	// TODO(samhassan): sendersCache should be an LRU capped cache
+	sendersCache map[common.Address][]common.Address
+	latestBlock  int64
+
+	authRcvr    authorized_receiver.AuthorizedReceiverInterface
+	offchainAgg offchain_aggregator_wrapper.OffchainAggregatorInterface
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	cacheMu sync.RWMutex
+	chStop  chan struct{}
+}
+
+func NewFwdMgr(db *sqlx.DB, cfg Config, client evmclient.Client, logpoller *evmlogpoller.LogPoller, lggr logger.Logger) *FwdMgr {
+	lggr = lggr.Named("FwdMgr")
+	lggr.Infow("Initializing EVM forwarder manager")
+	fwdMgr := FwdMgr{
+		logger:       lggr,
+		evmClient:    client,
+		ORM:          NewORM(db, lggr, cfg),
+		logpoller:    logpoller,
+		config:       cfg,
+		sendersCache: make(map[common.Address][]common.Address),
+		chStop:       make(chan struct{}),
+		cacheMu:      sync.RWMutex{},
+		latestBlock:  0,
+	}
+	return &fwdMgr
+}
+
+// Start starts the forwarder manager, init forwarder cache and listen to auth events for all forwarders
+func (f *FwdMgr) Start() error {
+	fwdrs, cnt, err := f.ORM.FindForwardersByChain(utils.Big(*f.evmClient.ChainID()))
+	if err != nil {
+		return errors.Errorf("Failed to retrieve forwarders for chain %d: %s", f.evmClient.ChainID(), err)
+	}
+	f.ctx, f.cancel = context.WithCancel(context.Background())
+	if cnt != 0 {
+		f.initForwardersCache(f.ctx, fwdrs)
+		f.subscribeForwardersLogs(fwdrs)
+	}
+
+	f.authRcvr, err = authorized_receiver.NewAuthorizedReceiver(common.Address{}, f.evmClient)
+	if err != nil {
+		f.logger.Criticalf("Failed to init AuthorizedReceiver")
+	}
+
+	f.offchainAgg, err = offchain_aggregator_wrapper.NewOffchainAggregator(common.Address{}, f.evmClient)
+	if err != nil {
+		f.logger.Criticalf("Failed to init OffchainAggregator")
+	}
+
+	go f.runLoop()
+	return nil
+}
+
+// TODO(samhassan): this should be aware of job type to decide how to fetch senders list.
+func (f *FwdMgr) MaybeForwardTransaction(from common.Address, to common.Address, EncodedPayload []byte) (fwdAddr common.Address, fwdPayload []byte, err error) {
+
+	senders, err := f.getContractSenders(to)
+	if err != nil {
+		return to, EncodedPayload, errors.Wrap(err, "Skipping forwarding transaction")
+	}
+
+	// Gets current forwarders that are in `to` senders
+	fwdrs, err := f.ORM.FindForwardersInListByChain(utils.Big(*f.evmClient.ChainID()), senders)
```

Comment:
Would it not be more efficient to just look these up in a map?

Reply:
Yes, but since we are maintaining a single cache, we are prune to cyclic behavior (or ub) when we look only in the map.